### PR TITLE
Add documentation for group search and app group API enhancement

### PR DESCRIPTION
**Pull request by chenningzhang-okta**
## Description:
- **What's changed?** 
  1. Added documentation for group search (Early Access feature)
  2. Added documentation for the new group attributes `source` and `source` link (Early Access feature. The two new group attributes are only present in groups with type `APP_GROUP`)
- **Is this PR related to a Monolith release?** Yes, 2020.10.0

### Resolves:

* [OKTA-325066](https://oktainc.atlassian.net/browse/OKTA-325066)

### Reviewers:
@joshreesjones-okta @daytonwilliams-okta @okta/ud-core

### CODE REVIEW

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md [hunk 1]
```diff
@@ -1838,6 +1981,7 @@ Specifies link relations. See [Web Linking](http://tools.ietf.org/html/rfc5988))
 | apps               | Lists all [applications](/docs/reference/api/apps/#application-object) that are assigned to the Group. See [Application Group Operations](/docs/reference/api/apps/#application-group-operations).          |
 | logo               | Provides links to logo images for the Group if available                     |
 | self               | The primary URL for the Group                                                                                             |
+| <ApiLifecycle access="ea" /> source             | The URL for the source [application](/docs/reference/api/apps/#application-object) of the group. This link attribute is only present in groups with `APP_GROUP` type.          |
```
Comment (chenningzhang-okta) [inline packages/@okta/vuepress-site/docs/reference/api/groups/index.md:1984]: @jakubvul-okta Can we add the `<ApiLifecycle access="ea" />` tag in a table? I want to indicate this individual attribute is an early access feature.

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md [hunk 2]
```diff
@@ -176,7 +177,8 @@ Enumerates Groups in your organization with pagination. A subset of Groups can b
 | after     | Specifies the pagination cursor for the next page of Groups                                | Query     | String   | FALSE    |         |
 | filter    | [Filter expression](/docs/reference/api-overview/#filtering) for Groups      | Query     | String   | FALSE    |         |
 | limit     | Specifies the number of Group results in a page                                            | Query     | Number   | FALSE    | 10000   |
-| q         | Searches the `name` property of Groups for matching value                                  | Query     | String   | FALSE    |         |
+| q         | Find groups that matches the `name` property                                               | Query     | String   | FALSE    |         |
+| <ApiLifecycle access="ea" /> search    | Searches for groups with a supported [filtering](/docs/reference/api-overview/#filtering) expression for most properties                        | Query     | String   | FALSE    |         |
```
Comment (joshreesjones-okta) [inline packages/@okta/vuepress-site/docs/reference/api/groups/index.md:181]: The "Early Access" label/span doesn't visually fit very well here in my opinion:

Screen Shot 2020-09-20 at 6 00 06 PM

Chenning or @jakubvul what are your thoughts on the positioning of this? Is there any way to make it fit on the same line (here and in other places on this page where it overflows)?

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md [hunk 3]
```diff
@@ -334,9 +344,9 @@ Link: <https://${yourOktaDomain}/api/v1/groups?after=00garwpuyxHaWOkdV0g4&limit=
 ]
 ```
 
-#### Search Groups
+#### Find Groups
 
-Searches for groups by `name` in your organization
+Find groups by `name` in your organization
```
Comment (joshreesjones-okta) [inline packages/@okta/vuepress-site/docs/reference/api/groups/index.md:349]: Nit: `Find` &rarr; `Finds`

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md [hunk 4]
```diff
@@ -732,6 +742,131 @@ Link: <https://${yourOktaDomain}/api/v1/groups?after=00g1emaKYZTWRYYRRTSK&limit=
 ]
 ```
 
+#### List Groups with Search
+
+<ApiLifecycle access="ea" />
+
+Searches for groups based on the properties specified in the search parameter
```
Comment (joshreesjones-okta) [inline packages/@okta/vuepress-site/docs/reference/api/groups/index.md:749]: Nit: `parameter` &rarr; `parameter.`

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md [hunk 5]
```diff
@@ -732,6 +742,131 @@ Link: <https://${yourOktaDomain}/api/v1/groups?after=00g1emaKYZTWRYYRRTSK&limit=
...
+
+```bash
+curl -v -X GET \
+-H "Accept: application/json" \
+-H "Content-Type: application/json" \
+-H "Authorization: SSWS ${api_token}" \
+"https://${yourOktaDomain}/api/v1/groups?search=lastUpdated+gt+\"2015-10-01T00:00:00.000Z\"+or+lastMembershipUpdated+gt+\"2015-10-01T00:00:00.000Z\""
+```
+
+##### Response example
```
Comment (joshreesjones-okta) [inline packages/@okta/vuepress-site/docs/reference/api/groups/index.md:790]: I don't think the two groups below match the filter above - can you update / run the example against a test org to make sure things are working as expected?

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md [hunk 6]
```diff
@@ -732,6 +742,131 @@ Link: <https://${yourOktaDomain}/api/v1/groups?after=00g1emaKYZTWRYYRRTSK&limit=
...
+
+This operation:
+
+* Supports [pagination](/docs/reference/api-overview/#pagination).
+* Requires [URL encoding](http://en.wikipedia.org/wiki/Percent-encoding).
+For example, `search=type eq "OKTA_GROUP"` is encoded as `search=type+eq+%22OKTA_GROUP%22`.
+Examples use cURL-style escaping instead of URL encoding to make them easier to read.
+Use an ID lookup for records that you update to ensure your results contain the latest data.
+* Searches many properties:
+   - Any group profile property, including imported app group profile properties.
```
Comment (joshreesjones-okta) [inline packages/@okta/vuepress-site/docs/reference/api/groups/index.md:761]: @donwoods-okta With the group profiles project, do we plan on supporting searching:
- any custom Okta group profile property, or
- any custom app group profile property that is added in the future?
cc @nghadge-okta 

Just want to make sure we don't accidentally promise too much on this line if we're not sure yet.

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md [hunk 7]
```diff
@@ -1733,6 +1868,14 @@ All groups have the following properties:
 
 > **Note:** The `id`, `created`, `lastUpdated`, `lastMembershipUpdated`, `objectClass`, `type`, and `_links` properties are available only after you create a Group.
 
+<ApiLifecycle access="ea" />
```
Comment (joshreesjones-okta) [inline packages/@okta/vuepress-site/docs/reference/api/groups/index.md:1871]: I'm wondering if we should move this to be next to `source` in the first column of the table. On the page, it seems like it could be a little ambiguous to some people whether it applies to the group attributes in the table above or not:

Screen Shot 2020-09-20 at 6 15 13 PM

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md (hunk 1 above)
Comment (joshreesjones-okta) [reply packages/@okta/vuepress-site/docs/reference/api/groups/index.md:1984]: I found that this is in fact used in tables on other pages (though it seems to fit a bit better on other pages).

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md [hunk 8]
```diff
@@ -250,6 +252,8 @@ curl -v -X GET \
 
 ##### Response example
 
+> **Note:** <ApiLifecycle access="ea" /> The `source` section and the `source` link in the response example below are only present in groups with type `APP_GROUP`. See [Group attributes](#group-attributes) and [Links object](#links-object).
```
Comment (joshreesjones-okta) [inline packages/@okta/vuepress-site/docs/reference/api/groups/index.md:255]: I have two (conflicting) thoughts:
1. Do we need this here, or can we leave the details about `source` in the "Group object" > "Group attributes" section below? (I think we can leave the details in the section below)
2. If we do want this, do we want to add details like this to other APIs that return groups? (I think no, that would be too verbose)

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md (hunk 4 above)
Comment (susanharper-okta) [reply packages/@okta/vuepress-site/docs/reference/api/groups/index.md:749]: this doesn't require a period as it isn't a complete sentence. Sentence fragments are the standard intro for endpoints, so it's okay to not have a period here.

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md (hunk 1 above)
Comment (susanharper-okta) [reply packages/@okta/vuepress-site/docs/reference/api/groups/index.md:1984]: > @jakubvul-okta Can we add the `<ApiLifecycle access="ea" />` tag in a table? I want to indicate this individual attribute is an early access feature.

We can and do add the that tag in tables, yes. And, yes, @joshreesjones-okta - sometimes formatting is problem with this label. You just have to play with, unfortunately.

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md (hunk 6 above)
Comment (donwoods-okta) [reply packages/@okta/vuepress-site/docs/reference/api/groups/index.md:761]: Certainly it is correct at this time: imported app group profile properties can be searched.  When we add support for custom properties in Okta groups, those properties will also be searchable.  We have not designed support for custom app group properties, but I expect those will also be searchable.

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md (hunk 2 above)
Comment (chenningzhang-okta) [reply packages/@okta/vuepress-site/docs/reference/api/groups/index.md:181]: It seems that the tag fits better if followed after the attribute name:
Screen Shot 2020-09-21 at 3 03 29 PM

Existing pattern on Apps page:
Screen Shot 2020-09-21 at 3 04 51 PM

Updated accordingly.

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md (hunk 5 above)
Comment (joshreesjones-okta) [reply packages/@okta/vuepress-site/docs/reference/api/groups/index.md:790]: The filter under Search Examples uses type eq "APP_GROUP" - was that intended?

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md (hunk 8 above)
Comment (joshreesjones-okta) [reply packages/@okta/vuepress-site/docs/reference/api/groups/index.md:255]: I was talking more about the note here (`> **Note:** <ApiLifecycle access="ea" /> The `source` section...`) - I agree `"source": ...` needs to be added to any `"APP_GROUP"` object in the documentation at GA time.

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md (hunk 5 above)
Comment (chenningzhang-okta) [reply packages/@okta/vuepress-site/docs/reference/api/groups/index.md:790]: The example listed under "Search Examples" is actually a different search than the one used in "request example". I tend to follow the same pattern as the existing User search documentation, where the "Search Examples" just gives example search strings and "Request Example" shows a different sample request and response.

Comment (joshreesjones-okta) [reply packages/@okta/vuepress-site/docs/reference/api/groups/index.md:790]: In that case, do we want to add a few more examples to further highlight that more things are searchable? Maybe:
```
source.id eq "blah"
profile.name eq "blah"
profile.<some AD group profile attribute> eq "blah"
<the existing compound filter from above>
```

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md (hunk 8 above)
Comment (chenningzhang-okta) [reply packages/@okta/vuepress-site/docs/reference/api/groups/index.md:255]: I do feel that it is necessary to add this note to point readers to the bottom of the page about the details of the attribute since it's a new attribute and is in early access stage.

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md (hunk 5 above)
Comment (chenningzhang-okta) [reply packages/@okta/vuepress-site/docs/reference/api/groups/index.md:794]: Added another compound search example -`Search Term Example` table displays simple filter examples and `Search Examples` lists compound search examples.

**Review by joshreesjones-okta — APPROVED** [review]

**Review by susanharper-okta — APPROVED** [review]
LGTM

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md [hunk 9]
```diff
@@ -176,7 +177,8 @@ Enumerates Groups in your organization with pagination. A subset of Groups can b
 | after     | Specifies the pagination cursor for the next page of Groups                                | Query     | String   | FALSE    |         |
 | filter    | [Filter expression](/docs/reference/api-overview/#filtering) for Groups      | Query     | String   | FALSE    |         |
 | limit     | Specifies the number of Group results in a page                                            | Query     | Number   | FALSE    | 10000   |
-| q         | Searches the `name` property of Groups for matching value                                  | Query     | String   | FALSE    |         |
+| q         | Find groups that matches the `name` property                                               | Query     | String   | FALSE    |         |
```
Comment (daytonwilliams-okta) [inline packages/@okta/vuepress-site/docs/reference/api/groups/index.md:180]: Nit: singular/plural agreement.  It's `groups that match` or `group that matches`.

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md [hunk 10]
```diff
@@ -176,7 +177,8 @@ Enumerates Groups in your organization with pagination. A subset of Groups can b
...
+| search <ApiLifecycle access="ea" /> | Searches for groups with a supported [filtering](/docs/reference/api-overview/#filtering) expression for most properties                        | Query     | String   | FALSE    |         |
```
Comment (daytonwilliams-okta) [inline packages/@okta/vuepress-site/docs/reference/api/groups/index.md:181]: "For most properties".  What properties would _not_ be included?  Do we have any more detail we could add to be a bit less vague?  E.g. if this is the same list that's enumerated below, maybe refer to it (or better yet, link to it)?

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md (hunk 8 above)
Comment (daytonwilliams-okta) [inline packages/@okta/vuepress-site/docs/reference/api/groups/index.md:255]: Nit: `groups of type APP_GROUP` (not "with type")

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md [hunk 11]
```diff
@@ -732,6 +742,135 @@ Link: <https://${yourOktaDomain}/api/v1/groups?after=00g1emaKYZTWRYYRRTSK&limit=
...
+
+Property names in the search parameter are case sensitive, whereas operators (`eq`, `sw`, etc.) and string values are case insensitive.
+
+This operation:
+
+* Supports [pagination](/docs/reference/api-overview/#pagination).
+* Requires [URL encoding](http://en.wikipedia.org/wiki/Percent-encoding).
+For example, `search=type eq "OKTA_GROUP"` is encoded as `search=type+eq+%22OKTA_GROUP%22`.
+Examples use cURL-style escaping instead of URL encoding to make them easier to read.
+Use an ID lookup for records that you update to ensure your results contain the latest data.
```
Comment (daytonwilliams-okta) [inline packages/@okta/vuepress-site/docs/reference/api/groups/index.md:759]: Maybe be a bit more explicit here & add a follow on sentence: `Search results are eventually consistent`.

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md [hunk 12]
```diff
@@ -732,6 +742,135 @@ Link: <https://${yourOktaDomain}/api/v1/groups?after=00g1emaKYZTWRYYRRTSK&limit=
...
+| `type eq "APP_GROUP"`                                     | Groups that have a `type` of `APP_GROUP`                                  |
+| `lastMembershipUpdated gt "yyyy-MM-dd'T'HH:mm:ss.SSSZ"`   | Groups whose memberships were last updated after a specific timestamp     |
+| `id eq "00gak46y5hydV6NdM0g4"`                            | Groups with a specified `id`                                              |
+| `profile.name eq "West Coast Users"`                      | Groups that have a `name` of `West Coast Users`                           |
+| `profile.samAccountName sw "West Coast" `                 | Groups whose `samAccountName` starts with `West Coast`                    |
+| `source.id` eq `0oa2v0el0gP90aqjJ0g7`                     | Groups that have the source application with a specified `source.id`      |
+
+##### Search Examples
+
+List groups with type `APP_GROUP` that were created before `01/01/2014` and have a source application id `0oa2v0el0gP90aqjJ0g7`.
```
Comment (daytonwilliams-okta) [inline packages/@okta/vuepress-site/docs/reference/api/groups/index.md:776]: Nit: `of` type APP_GROUP

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md [hunk 13]
```diff
@@ -732,6 +742,135 @@ Link: <https://${yourOktaDomain}/api/v1/groups?after=00g1emaKYZTWRYYRRTSK&limit=
...
+
+    search=type eq "APP_GROUP" and (created lt "2014-01-01T00:00:00.000Z" and source.id eq "0oa2v0el0gP90aqjJ0g7")
+
+List groups that have a `name` starts with `West Coast` or have the source application with id `0oa2v0el0gP90aqjJ0g7` or have a `samAccountName` of `West Coast Users`.
```
Comment (daytonwilliams-okta) [inline packages/@okta/vuepress-site/docs/reference/api/groups/index.md:780]: Nit: `that starts` or `starting`
Nit: `or whose source application has the id`

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md [hunk 14]
```diff
@@ -1733,6 +1872,12 @@ All groups have the following properties:
 
 > **Note:** The `id`, `created`, `lastUpdated`, `lastMembershipUpdated`, `objectClass`, `type`, and `_links` properties are available only after you create a Group.
 
+In addition, groups with type `APP_GROUP` also have the following properties:
+
+| Property                              | Description                                                                                     | DataType                   | Nullable | Unique | Readonly | MinLength | MaxLength | Validation |
+| ------------------------------------- | ----------------------------------------------------------------------------------------------- | -------------------------- | -------- | ------ | -------- | --------- | --------- | ---------- |
+| source <ApiLifecycle access="ea" />   | the ID of the source [application](/docs/reference/api/apps/#application-object) of the group.  | Array of String            | FALSE    | FALSE  | TRUE     |           |           |            |
```
Comment (daytonwilliams-okta) [inline packages/@okta/vuepress-site/docs/reference/api/groups/index.md:1879]: Nit: Capitalize `the` near `the ID of the source`

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md [hunk 15]
```diff
@@ -1838,6 +1983,7 @@ Specifies link relations. See [Web Linking](http://tools.ietf.org/html/rfc5988))
 | apps               | Lists all [applications](/docs/reference/api/apps/#application-object) that are assigned to the Group. See [Application Group Operations](/docs/reference/api/apps/#application-group-operations).          |
 | logo               | Provides links to logo images for the Group if available                     |
 | self               | The primary URL for the Group                                                                                             |
+| source <ApiLifecycle access="ea" /> | The URL for the source [application](/docs/reference/api/apps/#application-object) of the group. This link attribute is only present in groups with `APP_GROUP` type.          |
```
Comment (daytonwilliams-okta) [inline packages/@okta/vuepress-site/docs/reference/api/groups/index.md:1986]: Nit: `groups of APP_GROUP type`

**Review by daytonwilliams-okta — COMMENTED** [review]
Mostly LGTM, not quite ready to approve.

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md (hunk 10 above)
Comment (chenningzhang-okta) [reply packages/@okta/vuepress-site/docs/reference/api/groups/index.md:181]: I actually copied this description from Users search - https://developer.okta.com/docs/reference/api/users/#list-users. 

Looking at all group attributes (https://developer.okta.com/docs/reference/api/groups/#group-attributes), everything except for `_embedded`, `_links`, and `objectClass` is searchable. Would it be too specific if we put the description as `... for most [properties](#group-attributes)` except for `_embedded`, `_links`, and `objectClass``?

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md (hunk 8 above)
Comment (chenningzhang-okta) [reply packages/@okta/vuepress-site/docs/reference/api/groups/index.md:255]: updated here and everywhere else in this file

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md (hunk 10 above)
Comment (daytonwilliams-okta) [reply packages/@okta/vuepress-site/docs/reference/api/groups/index.md:181]: I'll defer to the docs experts on this one.  There's a complex tradeoff here between 'we have to update the docs every time we change something' and 'people can't figure out what exactly our product does'.

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md (hunk 14 above)
Comment (daytonwilliams-okta) [reply packages/@okta/vuepress-site/docs/reference/api/groups/index.md]: Well, the change a few lines down _does_ capitalize the `The`.  Is that part also consistent with the rest of its form?

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md [hunk 16]
```diff
@@ -732,14 +742,143 @@ Link: <https://${yourOktaDomain}/api/v1/groups?after=00g1emaKYZTWRYYRRTSK&limit=
...
+| `profile.samAccountName sw "West Coast" `                 | Groups whose `samAccountName` starts with `West Coast`                    |
+| `source.id` eq `0oa2v0el0gP90aqjJ0g7`                     | Groups that have the source application with a specified `source.id`      |
+
+##### Search Examples
+
+List groups of type `APP_GROUP` that were created before `01/01/2014` and have a source application id `0oa2v0el0gP90aqjJ0g7`.
+
+    search=type eq "APP_GROUP" and (created lt "2014-01-01T00:00:00.000Z" and source.id eq "0oa2v0el0gP90aqjJ0g7")
+
+List groups that have a `name` that starts with `West Coast` or whose source application has the id `0oa2v0el0gP90aqjJ0g7` or have a `samAccountName` of `West Coast Users`.
```
Comment (daytonwilliams-okta) [inline packages/@okta/vuepress-site/docs/reference/api/groups/index.md:780]: Hm, looking more carefully, there's a devious ambiguity here.  This can be read 2 ways:

1: List groups that ``have a `name` that starts with `West Coast` `` or ``whose source application has the id `0oa2v0el0gP90aqjJ0g7` `` or ``have a `samAccountName` of `West Coast Users`.``

2: List groups that ``have a `name` that starts with `West Coast` `` or whose source application (``has the id `0oa2v0el0gP90aqjJ0g7` or have a `samAccountName` of `West Coast Users` ``).

In other words, grouping the third clause as a sub-clause of `or whose source application` or as a subclause of `list groups that`.  The second interpretation is flat out wrong.  `samAccountName` has nothing to do with the source application

I'm not sure how to rewrite this to avoid the ambiguity, but I don't think we should leave it like this.

Comment (chenningzhang-okta) [reply packages/@okta/vuepress-site/docs/reference/api/groups/index.md:780]: hmm what about:
`List groups that have a `name` that starts with `West Coast` or have a `samAccountName` of `West Coast Users` or whose source application has the id `0oa2v0el0gP90aqjJ0g7``?

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md (hunk 10 above)
Comment (susanharper-okta) [reply packages/@okta/vuepress-site/docs/reference/api/groups/index.md:181]: I'm leaning towards giving the people what they need to figure out what exactly our product does and dealing with updating the docs. Updating our docs is fairly easy (if Jiras are created for said changes in a timely manner).

File: packages/@okta/vuepress-site/docs/reference/api/groups/index.md (hunk 14 above)
Comment (susanharper-okta) [reply packages/@okta/vuepress-site/docs/reference/api/groups/index.md]: @daytonwilliams-okta and @chenningzhang-okta: entries in tables are supposed to be sentence fragments, which don't require a period. I would start these fragments with a capital letter.

The only time we have periods is if there are a mix of sentence fragments and complete sentences provided in these descriptions.

**Review by daytonwilliams-okta — APPROVED** [review]
LGTM, :rocket: :sheep: :it:
